README.md
**A flexible Flutter widget that supports the PNG and SVG images from the network, can cache them. Actually 3 in 1, supports PNG SVG as well as image from local assets**

<div align="center">
    <img src="https://raw.githubusercontent.com/juni12891226/flexible-png-svg-network-image-widget-juni1289/main/Screenshot_1701360239.png" width="300px" height="600px">  
</div>

## Features

* Supports PNG Network Image with Cache
* Supports SVG Network Image with Cache
* Supports Images files from Local Assets

* You just have to select if you have to load the image from the network or from the local assets using **bool -> isNetworkImage**
* You just have to provide the **String -> imagePathOrURL** to specify the path of the image
* You can cache PNG Images as well as SVG Images as long as you want!

## Additional Information

* The package itself checks for the PNG or SVG extension
* Easy to manage and flexible to use

## Basic Setup

```
    FlexiblePngSvgNetworkImageWidget(
              networkErrorPlaceHolderImagePath: placeHolderImagePathLocal,
              imagePathOrURL: imageURL,
              isNetworkImage: true | false,
              height: 50,
              width: 30,
              filterQuality: FilterQuality.high,
              svgLoadRetryLimit: 2,
                cacheRuleStaleDuration:Duration(days: 100)
            )
```

## Main Widget Signature

```
    ///[imagePathOrURL] is required | image path from the assets or the network URL
    ///[height] is required
    ///[width] is required
    ///[boxFit] is optional
    ///[filterQuality] is optional and default is High Quality
    ///[colorSVG] is optional
    ///[isNetworkImage] is required | true if load from network | false if load from assets
    ///[colorProgressLoaderIndicator] is optional default is blue
    ///[networkErrorPlaceHolderImagePath] is required | in case of error the default placeholder image asset path
    ///[errorWidget] is optional | if you want to show your own error widget in case of png image
    ///[progressIndicatorWidget] is optional | if you want to show your own progress widget
    ///[networkTimeOutDuration] is optional | for SVG you can set the load timeout | default is 30 seconds
    ///[svgLoadRetryLimit] is optional | for SVG you can set the retry load limit | default is 1 times | single
    ///[useDiskCacheForSVG] is optional | default is true
    ///[cacheRuleStaleDuration] is optional | for how much time the images needs to be caches | default is 500 days
    ///[onSVGLoadFailedCallback] is optional
    ///[useOldPngImageOnUrlChange] in order to use the old image on new URL default is false
    const FlexiblePngSvgNetworkImageWidget(
        {Key? key,
        required this.networkErrorPlaceHolderImagePath,
        required this.imagePathOrURL,
        required this.height,
        required this.width,
        this.useOldPngImageOnUrlChange = false,
        this.networkTimeOutDuration,
        this.useDiskCacheForSVG,
        this.cacheRuleStaleDuration,
        this.onSVGLoadFailedCallback,
        this.svgLoadRetryLimit,
        this.progressIndicatorWidget,
        this.errorWidget,
        this.colorProgressLoaderIndicator,
        this.isNetworkImage = false,
        this.colorSVG,
        this.boxFit,
        this.filterQuality})
            : super(key: key);
```

CHANGELOG.md
## 0.0.1

* Initial Release

## 0.0.2

* Example added
* Performance Improvements
* Documentation Updated

## 0.0.3

* Example Updated
* Readme updated
* Performance Improvements
* Basic Usage Added

## 0.0.4

* Screen shot added

## 0.0.5

* SVG network image widget added with new callback
* Documentation added
* Main Widget Class name refactored
* Performance Improvements


example/lib/main.dart
import 'package:flexible_png_svg_network_image_widget_juni1289/flexible_png_svg_network_image_juni1289.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String pngURL = "https://openexpoeurope.com/wp-content/uploads/2019/12/flutter-logo-sharing.png";
  String svgURL = "https://raw.githubusercontent.com/dart-lang/site-shared/master/src/_assets/image/flutter/logo/default.svg";

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Example"),
      ),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.center,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: <Widget>[
            const Text(
              "from Assets with error",
              textAlign: TextAlign.center,
            ),
            const FlexiblePngSvgNetworkImageWidget(
              networkErrorPlaceHolderImagePath: "assets/images/placeholder.png",
              imagePathOrURL: "assets/icons/cat.png",
              height: 50,
              width: 30,
              filterQuality: FilterQuality.high,
            ),
            const Text(
              "from Assets with no error",
              textAlign: TextAlign.center,
            ),
            const FlexiblePngSvgNetworkImageWidget(
              networkErrorPlaceHolderImagePath: "assets/images/placeholder.png",
              imagePathOrURL: "assets/images/cat.png",
              height: 50,
              width: 30,
              filterQuality: FilterQuality.high,
            ),
            const Text(
              "from network for PNG with no error",
              textAlign: TextAlign.center,
            ),
            FlexiblePngSvgNetworkImageWidget(
              networkErrorPlaceHolderImagePath: "assets/images/placeholder.png",
              imagePathOrURL: pngURL,
              height: 90,
              width: 60,
              filterQuality: FilterQuality.high,
              isNetworkImage: true,
              colorProgressLoaderIndicator: Colors.purple,
            ),
            const Text(
              "from network for PNG with error",
              textAlign: TextAlign.center,
            ),
            const FlexiblePngSvgNetworkImageWidget(
              networkErrorPlaceHolderImagePath: "assets/images/placeholder.png",
              imagePathOrURL: "https://openexpoeurope.com/wp-content/uploads/2019/12.png",
              height: 90,
              width: 60,
              filterQuality: FilterQuality.high,
              isNetworkImage: true,
              colorProgressLoaderIndicator: Colors.purple,
            ),
            const Text(
              "from network for SVG with no error",
              textAlign: TextAlign.center,
            ),
            SizedBox(
              height: 100,
              width: 100,
              child: Center(
                child: FlexiblePngSvgNetworkImageWidget(
                  networkErrorPlaceHolderImagePath: "assets/images/placeholder.png",
                  imagePathOrURL: svgURL,
                  height: 90,
                  width: 60,
                  filterQuality: FilterQuality.high,
                  isNetworkImage: true,
                  colorProgressLoaderIndicator: Colors.purple,
                ),
              ),
            ),
            const Text(
              "from network for SVG with error",
              textAlign: TextAlign.center,
            ),
            SizedBox(
              height: 100,
              width: 100,
              child: Center(
                child: FlexiblePngSvgNetworkImageWidget(
                  networkErrorPlaceHolderImagePath: "assets/images/placeholder.png",
                  imagePathOrURL: "${svgURL}sdkjk",
                  height: 90,
                  width: 60,
                  filterQuality: FilterQuality.high,
                  isNetworkImage: true,
                  colorProgressLoaderIndicator: Colors.purple,
                ),
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/png_cached_network_image_view.dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flexible_png_svg_network_image_widget_juni1289/flexible_png_svg_network_image_juni1289.dart';
import 'package:flutter/material.dart';

///PNG cached network image
///To load the PNG images from the URL
///And cache them on the device storage until the image loader URL gets updated
class PNGCachedNetworkImageView extends StatelessWidget {
  final String imagePathOrURL;
  final double height;
  final double width;
  final BoxFit? boxFit;
  final FilterQuality? filterQuality;
  final Color? colorProgressLoaderIndicator;
  final String networkErrorPlaceHolderImagePath;
  final Widget? errorWidget;
  final Widget? progressIndicatorWidget;
  final bool useOldPngImageOnUrlChange;

  ///[imagePathOrURL] is required | image path from the assets or the network URL
  ///[height] is required
  ///[width] is required
  ///[boxFit] is optional
  ///[filterQuality] is optional and default is High Quality
  ///[colorProgressLoaderIndicator] is optional default is blue
  ///[networkErrorPlaceHolderImagePath] is required | in case of error the default placeholder image asset path
  ///[errorWidget] is optional | if you want to show your own error widget in case of png image
  ///[progressIndicatorWidget] is optional | if you want to show your own progress widget
  ///[useOldPngImageOnUrlChange] in order to use the old image on new URL default is false
  const PNGCachedNetworkImageView(
      {Key? key,
      this.useOldPngImageOnUrlChange = false,
      this.progressIndicatorWidget,
      required this.networkErrorPlaceHolderImagePath,
      this.colorProgressLoaderIndicator,
      this.errorWidget,
      this.boxFit,
      required this.imagePathOrURL,
      required this.height,
      required this.width,
      this.filterQuality})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return boxFit != null
        ? CachedNetworkImage(
            useOldImageOnUrlChange: useOldPngImageOnUrlChange,
            imageUrl: imagePathOrURL,
            height: height,
            width: width,
            filterQuality: filterQuality ?? FilterQuality.high,
            fit: boxFit,
            progressIndicatorBuilder:
                (BuildContext context, url, downloadProgress) {
              return progressIndicatorWidget ??
                  SizedBox(
                      height: height,
                      width: width,
                      child: Center(
                          child: CircularProgressIndicator(
                              color: colorProgressLoaderIndicator ??
                                  Colors.blue)));
            },
            errorWidget: (BuildContext context, url, error) {
              return errorWidget ??
                  Center(
                      child: FlexiblePngSvgNetworkImageWidget(
                          networkErrorPlaceHolderImagePath:
                              networkErrorPlaceHolderImagePath,
                          imagePathOrURL: networkErrorPlaceHolderImagePath,
                          height: height,
                          width: width,
                          filterQuality: filterQuality ?? FilterQuality.high));
            })
        : CachedNetworkImage(
            imageUrl: imagePathOrURL,
            height: height,
            width: width,
            filterQuality: filterQuality ?? FilterQuality.high,
            progressIndicatorBuilder:
                (BuildContext context, url, downloadProgress) {
              return progressIndicatorWidget ??
                  SizedBox(
                      height: height,
                      width: width,
                      child: Center(
                          child: CircularProgressIndicator(
                              color: colorProgressLoaderIndicator ??
                                  Colors.blue)));
            },
            errorWidget: (BuildContext context, url, error) {
              return errorWidget ??
                  Center(
                      child: FlexiblePngSvgNetworkImageWidget(
                          networkErrorPlaceHolderImagePath:
                              networkErrorPlaceHolderImagePath,
                          imagePathOrURL: networkErrorPlaceHolderImagePath,
                          height: height,
                          width: width,
                          filterQuality: filterQuality ?? FilterQuality.high));
            });
  }
}


lib/svg_cached_network_image_view.dart
import 'dart:typed_data';

import 'package:flutter/material.dart';
import 'package:flutter_advanced_networkimage_2/provider.dart';
import 'package:flutter_svg/flutter_svg.dart';

///SVG cached network image
///To load the SVG images from the URL
///And cache them on the device storage until the image loader URL gets updated
///The [svgCacheRuleStaleDuration] can be provided for cache stale period
class SVGCachedNetworkImageView extends StatelessWidget {
  ///The image path URL for the SVG image to be loaded from the remote server
  final String imagePathOrURL;

  ///To set the SVG retry load limit, default is 1
  final int? svgLoadRetryLimit;

  ///The duration for setting the timeout from the remote server
  final Duration? timeOutDuration;

  ///The flag control the disk cache will be used or not.
  final bool? useDiskCacheForSVG;

  ///This duration can be used to set the cache stale rule
  ///The default duration is 500 days
  final Duration? svgCacheRuleStaleDuration;

  ///The callback is for checking when the SVG is failed to load
  final Function? onSVGLoadFailedCallback;

  ///The callback can be used to check when the SVG is loaded
  final Function? svgLoadedCallback;

  ///[imagePathOrURL] is required
  ///[timeOutDuration] is optional | for SVG you can set the load timeout | default is 30 seconds
  ///[svgLoadRetryLimit] is optional | for SVG you can set the retry load limit | default is 1 times | single
  ///[useDiskCacheForSVG] is optional | default is true
  ///[svgCacheRuleStaleDuration] is optional | for how much time the images needs to be caches | default is 500 days
  ///[onSVGLoadFailedCallback] is optional
  const SVGCachedNetworkImageView(
      {Key? key,
      this.svgLoadedCallback,
      this.timeOutDuration,
      this.useDiskCacheForSVG,
      this.svgCacheRuleStaleDuration,
      this.onSVGLoadFailedCallback,
      this.svgLoadRetryLimit,
      required this.imagePathOrURL})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return SvgPicture(AdvancedNetworkSvg(
        imagePathOrURL,
        (theme) => (bytes, colorFilter, key) {
              return svg.svgPictureDecoder(
                  bytes ?? Uint8List.fromList(const []),
                  false,
                  colorFilter,
                  key,
                  theme: theme);
            },
        retryLimit: svgLoadRetryLimit ?? 1,
        timeoutDuration: timeOutDuration ?? const Duration(seconds: 30),
        useDiskCache: useDiskCacheForSVG ?? true,
        cacheRule: CacheRule(
            maxAge: svgCacheRuleStaleDuration ?? const Duration(days: 500),
            storeDirectory: StoreDirectoryType.temporary),
        loadFailedCallback: () {
      if (onSVGLoadFailedCallback != null) {
        onSVGLoadFailedCallback!();
      }
    }, loadedCallback: () {
      if (svgLoadedCallback != null) {
        svgLoadedCallback!();
      }
    }));
  }
}


lib/flexible_png_svg_network_image_juni1289.dart
import 'package:flexible_png_svg_network_image_widget_juni1289/empty_container_helper_widget.dart';
import 'package:flexible_png_svg_network_image_widget_juni1289/png_cached_network_image_view.dart';
import 'package:flexible_png_svg_network_image_widget_juni1289/svg_cached_network_image_view.dart';
import 'package:flutter/material.dart';
import 'package:flutter_svg/flutter_svg.dart';

///A Flutter widget to load the images from the network or from the local assets
///It can load the images from the URL for PNG as well as SVG images
///It decided for the image file extension itself
///The [isNetworkImage] param decides whether to load the image from the URL or from the local assets
///Mainly there are 2 image file types can be loaded PNG or SVG
class FlexiblePngSvgNetworkImageWidget extends StatelessWidget {
  ///The image path or the URL for the image to be loaded
  ///Determines the type of the image PNG or SVG
  final String imagePathOrURL;

  ///Required param to set the height of the image
  final double height;

  ///Required param to set the width of the image
  final double width;

  ///To set the boxfit props for the image
  final BoxFit? boxFit;

  ///To set the filter quality for the PNG images, default is High
  final FilterQuality? filterQuality;

  ///To set the SVG color, depends if your provided SVG supports the color change
  final Color? colorSVG;

  ///To set if this widget should load the image from the remote server or from the load assets
  final bool isNetworkImage;

  ///To set the color for the progress loader for PNG images
  ///By default a progress loader with blue color would be shown
  ///You can provide your own widget while waiting for the PNG image to load
  final Color? colorProgressLoaderIndicator;

  ///To set the path of the network image placeholder that is when the image fails to load and this as a default image to be shown
  final String networkErrorPlaceHolderImagePath;

  ///To set the error widget
  ///Custom widget that you can set in place of as an error widget
  ///When the image is failed to load
  final Widget? errorWidget;

  ///If you want to show your own progress indicator widget
  ///You can use Lottie animation widget
  ///By default a progress widget is shown
  final Widget? progressIndicatorWidget;

  ///To set the SVG load retry limit
  ///The default is set to 1
  final int? svgLoadRetryLimit;

  ///To set the network timeout duration while loading the image from the remote network
  ///The default is 30 seconds
  final Duration? networkTimeOutDuration;

  ///The flag control the disk cache will be used or not.
  final bool? useDiskCacheForSVG;

  ///This duration can be used to set the cache stale rule
  ///The default duration is 500 days
  final Duration? svgCacheRuleStaleDuration;

  ///The callback is for checking when the SVG is failed to load
  final Function? onSVGLoadFailedCallback;

  ///To set if it should use the old PNG image for a new URL
  ///if it is set to true, will always use the old image for every new URL
  ///else, will always change the image for every new URL
  ///Cache refreshing
  ///Default value is false
  final bool useOldPngImageOnUrlChange;

  ///The callback can be used to check when the SVG is loaded
  final Function? svgLoadedCallback;

  ///A generic widget to show the images from the Assets or from the network
  ///[imagePathOrURL] is required | image path from the assets or the network URL
  ///[height] is required
  ///[width] is required
  ///[boxFit] is optional
  ///[filterQuality] is optional and default is High Quality
  ///[colorSVG] is optional
  ///[isNetworkImage] is required | true if load from network | false if load from assets
  ///[colorProgressLoaderIndicator] is optional default is blue
  ///[networkErrorPlaceHolderImagePath] is required | in case of error the default placeholder image asset path
  ///[errorWidget] is optional | if you want to show your own error widget in case of png image
  ///[progressIndicatorWidget] is optional | if you want to show your own progress widget
  ///[networkTimeOutDuration] is optional | for SVG you can set the load timeout | default is 30 seconds
  ///[svgLoadRetryLimit] is optional | for SVG you can set the retry load limit | default is 1 times | single
  ///[useDiskCacheForSVG] is optional | default is true
  ///[svgCacheRuleStaleDuration] is optional | for how much time the images needs to be caches | default is 500 days
  ///[onSVGLoadFailedCallback] is optional
  ///[useOldPngImageOnUrlChange] in order to use the old image on new URL default is false
  const FlexiblePngSvgNetworkImageWidget(
      {Key? key,
      required this.networkErrorPlaceHolderImagePath,
      required this.imagePathOrURL,
      required this.height,
      required this.width,
      this.svgLoadedCallback,
      this.useOldPngImageOnUrlChange = false,
      this.networkTimeOutDuration,
      this.useDiskCacheForSVG,
      this.svgCacheRuleStaleDuration,
      this.onSVGLoadFailedCallback,
      this.svgLoadRetryLimit,
      this.progressIndicatorWidget,
      this.errorWidget,
      this.colorProgressLoaderIndicator,
      this.isNetworkImage = false,
      this.colorSVG,
      this.boxFit,
      this.filterQuality})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    //check if load from network
    if (isNetworkImage) {
      //check if the url is for SVG
      if (imagePathOrURL.endsWith(".svg")) {
        return SVGCachedNetworkImageView(
          svgLoadedCallback: svgLoadedCallback,
          imagePathOrURL: imagePathOrURL,
          timeOutDuration: networkTimeOutDuration,
          useDiskCacheForSVG: useDiskCacheForSVG,
          svgCacheRuleStaleDuration: svgCacheRuleStaleDuration,
          onSVGLoadFailedCallback: onSVGLoadFailedCallback,
          svgLoadRetryLimit: svgLoadRetryLimit,
        );
      } else {
        //url is for PNG
        return PNGCachedNetworkImageView(
            useOldPngImageOnUrlChange: useOldPngImageOnUrlChange,
            progressIndicatorWidget: progressIndicatorWidget,
            errorWidget: errorWidget,
            networkErrorPlaceHolderImagePath: networkErrorPlaceHolderImagePath,
            colorProgressLoaderIndicator: colorProgressLoaderIndicator,
            boxFit: boxFit,
            imagePathOrURL: imagePathOrURL,
            height: height,
            width: width,
            filterQuality: filterQuality);
      }
    } else {
      //load from assets
      return _getImageFromAssetView();
    }
  }

  ///Load the image from the local assets
  Widget _getImageFromAssetView() {
    if (imagePathOrURL.isNotEmpty) {
      if (boxFit != null) {
        if (imagePathOrURL.endsWith(".png")) {
          return Image.asset(imagePathOrURL,
              height: height,
              width: width,
              filterQuality: filterQuality ?? FilterQuality.high,
              fit: boxFit, errorBuilder:
                  (BuildContext context, Object error, StackTrace? stackTrace) {
            return errorWidget ??
                Image.asset(networkErrorPlaceHolderImagePath,
                    height: height,
                    width: width,
                    filterQuality: filterQuality ?? FilterQuality.high);
          });
        } else if (imagePathOrURL.endsWith(".svg")) {
          return SvgPicture.asset(imagePathOrURL,
              height: height, width: width, fit: boxFit!, color: colorSVG);
        } else {
          return const EmptyContainerHelperWidget();
        }
      } else {
        if (imagePathOrURL.endsWith(".png")) {
          return Image.asset(imagePathOrURL,
              height: height,
              width: width,
              filterQuality: filterQuality ?? FilterQuality.high, errorBuilder:
                  (BuildContext context, Object error, StackTrace? stackTrace) {
            return errorWidget ??
                Image.asset(networkErrorPlaceHolderImagePath,
                    height: height,
                    width: width,
                    filterQuality: filterQuality ?? FilterQuality.high);
          });
        } else if (imagePathOrURL.endsWith(".svg")) {
          return SvgPicture.asset(imagePathOrURL,
              height: height, width: width, color: colorSVG);
        } else {
          return const EmptyContainerHelperWidget();
        }
      }
    } else {
      return const EmptyContainerHelperWidget();
    }
  }
}
